nucypher/policy/conditions/address.py

```diff
@@ -0,0 +1,118 @@
+from typing import Dict, List, Optional, Tuple, Union
+
+from eth_utils import to_checksum_address
+from marshmallow import (
+    ValidationError,
+    fields,
+    post_load,
+    validate,
+    validates,
+)
+
+from nucypher.policy.conditions.base import AccessControlCondition
+from nucypher.policy.conditions.context import (
+    USER_ADDRESS_CONTEXT,
+    resolve_any_context_variables,
+)
+from nucypher.policy.conditions.exceptions import (
+    InvalidConditionContext,
+)
+from nucypher.policy.conditions.lingo import (
+    ConditionType,
+)
+
+# Maximum number of addresses allowed in the address allowlist
+MAX_ALLOWLIST_ADDRESSES = 25
+
+
+class AddressAllowlistCondition(AccessControlCondition):
+    """
+    A condition that checks if a user's wallet address is in a list of allowed addresses.
+    The user must provide a signed message to prove ownership of the wallet.
+    """
+
+    CONDITION_TYPE = ConditionType.ADDRESS_ALLOWLIST.value
+
+    class Schema(AccessControlCondition.Schema):
+        condition_type = fields.Str(
+            validate=validate.Equal(ConditionType.ADDRESS_ALLOWLIST.value),
+            required=True,
+        )
+        user_address = fields.Str(
+            validate=validate.Equal(USER_ADDRESS_CONTEXT),
+            required=True,
+        )
+        addresses = fields.List(
+            fields.String(required=True),
+            required=True,
+            validate=validate.Length(min=1, max=MAX_ALLOWLIST_ADDRESSES),
+        )
+
+        @validates("addresses")
+        def validate_addresses(self, addresses):
+            # Check for duplicates
+            if len(set(addresses)) != len(addresses):
+                raise ValidationError("Duplicate addresses are not allowed")
+
+            # Validate all addresses
+            for address in addresses:
+                try:
+                    # Ensure addresses have proper checksum
+                    normalized_address = to_checksum_address(address)
+                except ValueError as e:
+                    raise ValidationError(
+                        f"Invalid Ethereum address: '{address}', provided in the list of addresses"
+                    ) from e
+                if normalized_address != address:
+                    raise ValidationError(
+                        f"Address '{address}', provided in the list of addresses, is not a checksummed address"
+                    )
+
+        @post_load
+        def make(self, data, **kwargs):
+            return AddressAllowlistCondition(**data)
+
+    def __init__(
+        self,
+        user_address: str,
+        addresses: List[str],
+        name: Optional[str] = None,
+        condition_type: str = ConditionType.ADDRESS_ALLOWLIST.value,
+        *args,
+        **kwargs,
+    ):
+        self.user_address = user_address
```

Comment:
If `user_address` is supposed to be equal than `USER_ADDRESS_CONTEXT`, shouldn't we initialize it directly in the constructor, by default? We do something similar in other condition schemas.

Reply:
Actually the usage of this variable as an explicit provided one was done according to: https://github.com/nucypher/nucypher/pull/3601#discussion_r2138442235. And I took `jwt.py` as example. Please, let me know if I did not get it right.

Reply:
That case is different since the `jwtToken` value can be any user-defined context variable (`:jwtToken` or `:token` or ...) - what it is providing is just a key-value pair.

In this case you have to use a specific context variable `:userAddress` so that the logic for ensuring the user provides a signed message to prove ownership of the wallet address is adhered to.

Reply:
Thanks @derekpierre 
I meat I took the approach of jwtToken by having an explicit variable as it was suggested to use explicit and not implicit. The variable in user_address is always `:userAddress` so I enforced that by throwing if it was not.  What is the exact point to be done in this regard?
I think @cygnusv was recommending using implicit value by not passing a value it to the constructor. Or is this something else meant?

Reply:
What @cygnusv meant, is using ":userAddress" as the default value in the python constructor - my suggestion here https://github.com/nucypher/nucypher/pull/3601#discussion_r2163762093 does that. That way you don't have to always pass the `user_address` parameter in your tests. It's the same philosophy we use for condition type in constructors.

Reply:
Seems to be resolved via - https://github.com/nucypher/nucypher/pull/3601/commits/d53b641353d5e791bc9d1151f8676af42a7effb6.